### ISSUE TITLE
Flow analysis spec and implementation discrepancy around never assigned variables

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#1138]
The flow analysis implementation seems to use some global knowledge about what variables are never assigned that isn't captured in the current spec.  The behavior doesn't seem incorrect to me, but needs to be captured in the specification.  Specifically, consider this program:

```dart
void use(x) {}
test(bool b) {
  {
    late int y;
    var f = () => y;
  }
  {
    late int y;
    var f = () => y;
    y = 3;
  }
}
```

The front ends report an error on the use of `y` in the first closure, but not the second.  Presumably this is based on the fact that `y` is never assigned at all in the first example?  This seems reasonable, but is not currently part of the specified behavior.  What exactly is the implementation tracking here?

cc @stereotype441 @johnniwinther @scheglov

**Comment by stereotype441** [comment]
You're right, this is not specified--it falls under the TODO comment "TODO: Add missing expressions, handle cascades and left-hand sides accurately".  The rule that's currently implemented looks something like this:

If `N` is a local function or closure of the form `() B` (where `B` is a function body), then:
- Let `capturedHere` be the set of variables that are write captured by `N` (meaning that they are written to within `B` but declared outside `B`).
- Let `writtenAnywhere` be the set of variables that are assigned to anywhere in the body of the top-level function being analyzed.
- Let `capturedAnywhere` be the set of variables that are write captured by a local function or closure anywhere within the body of the top-level function being analyzed.
- Let `before(B) = conservativeJoin(before(N), writtenAnywhere, capturedAnywhere)`
- Let `after(N) = conservativeJoin(before(N), {}, capturedHere)`

(The reasoning is since the closure might be called at any point in the future, we have to assume that if a variable is written to anywhere, it might get written to before the closure body executes; and similarly, if a variable is captured anywhere, it might get captured before the closure body executes).
